File: devices/riccardo/lkp_enc_driver.c
#include <linux/ioctl.h>
#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/fs.h>
#include <linux/uaccess.h>
#include <linux/io.h>
#include <linux/slab.h>
#include <linux/types.h>
#include <linux/errno.h>

#define LKP_ENC_WRITE_SEED _IOW('q', 1, unsigned int)
#define LKP_ENC_READ_SEED _IOR('q', 2, unsigned int)

#define LKP_ENC_WRITE_STRING _IOW('q', 3, char *)
#define LKP_ENC_READ_STRING _IOR('q', 4, char *)

#define DEVICE_BASE_PHYS_ADDR 0xfebd1000

void *devmem = 0x0;

static long lkp_enc_ioctl(struct file *file, unsigned int cmd, unsigned long arg) {
    int seed, i;
    char *string;

    switch (cmd) {
        case LKP_ENC_WRITE_SEED:
            if (copy_from_user(&seed, (int *)arg, sizeof(int)) != 0)
                return -EFAULT; // Failed to copy seed from user space
            iowrite32(seed, devmem);
            break;

        case LKP_ENC_READ_SEED:
            seed = ioread32(devmem);
            if (copy_to_user((int *)arg, &seed, sizeof(int)) != 0)
                return -EFAULT; // Failed to copy seed to user space
            break;

        case LKP_ENC_WRITE_STRING:
            string = kmalloc(256, GFP_KERNEL);
            if (!string)
                return -ENOMEM; // Failed to allocate memory
                
            if(copy_from_user(string, (char *)arg, 256) != 0) {
                kfree(string);
                return -EFAULT; // Failed to copy string from user space
            }

            for (i = 0; i < 255 && string[i] != '\0'; i++) {
                iowrite8(string[i], devmem + 4 + i);
            }
            iowrite8('\0', devmem + 4 + i);

            kfree(string);
            break;

        case LKP_ENC_READ_STRING:
            string = kmalloc(256, GFP_KERNEL);

            if(!string) return -ENOMEM; // Failed to allocate memory

            do{
                string[i] = ioread8(devmem + 4 + i);
                i++;
            } while(i < 255 && string[i] != '\0');

            string[i] = '\0'; // Null-terminate the string

            if(copy_to_user((char *)arg, string, 256) != 0) {
                kfree(string);
                return -EFAULT;
            }
            
            break;
        default:
            return -ENOTTY; // Unknown command
    }

    return 0; // Success
}


static struct file_operations lkp_enc_fops = {
    .unlocked_ioctl = lkp_enc_ioctl,
};

/**
 * Use copy_to_user and copy_from_user
*/

static int __init lkp_enc_driver_init(void) {
    devmem = ioremap(DEVICE_BASE_PHYS_ADDR, 4096);

    if(!devmem) {
        printk(KERN_ERR "Failed to map device registers in memory");
        return -1;
    }

    if (register_chrdev(250, "lkp_enc", &lkp_enc_fops) < 0) {
        printk(KERN_ERR "Failed to register lkp_enc_driver\n");
        return -1;
    }

    pr_info("lkp_enc_driver loaded\n"
            "Registered ioctl commands:\n"
            "0x%lx: LKP_ENC_WRITE_SEED\n"
            "0x%lx: LKP_ENC_READ_SEED\n"
            "0x%lx: LKP_ENC_WRITE_STRING\n"
            "0x%lx: LKP_ENC_READ_STRING\n", 
            LKP_ENC_WRITE_SEED, LKP_ENC_READ_SEED, 
            LKP_ENC_WRITE_STRING, LKP_ENC_READ_STRING);

    return 0;
}

static void __exit lkp_enc_driver_exit(void) {
    unregister_chrdev(250, "my_rng_driver");

    if(devmem)
        iounmap(devmem);

    printk(KERN_INFO "lkp_enc_driver unloaded\n");
}

module_init(lkp_enc_driver_init);
module_exit(lkp_enc_driver_exit);

File: devices/riccardo/lkp_enc_driver.rs
//! Rust HW9 driver module
//! 
//! This module is a Rust implementation of the HW9 driver module written in C.

use core::result::Result::Ok;
use core::ffi::c_void;
use kernel::sync::Mutex;
use kernel::bindings;
use kernel::user_ptr::UserSlicePtr;
// , _IOC_SIZESHIFT, _IOC_SIZEMASK
use kernel::{
    ioctl::{_IOR, _IOW}, prelude::*, Module
};
use kernel::file;
use kernel::chrdev;

module! {
    type: LkpEnc,
    name: "lkp_enc",
    author: "Riccardo Strina",
    description: "HW9... but in Rust!",
    license: "GPL",
}

const LKP_ENC_WRITE_SEED: u32 = _IOW::<u32>('q' as u32, 1);
const LKP_ENC_WRITE_STRING: u32 = _IOW::<&str>('q' as u32, 2);
const LKP_ENC_READ_STRING: u32 = _IOR::<&str>('q' as u32, 3);

static DEVMEM: Mutex<usize> = unsafe {
    Mutex::new(0)
};

#[allow(dead_code)]
const DEVICE_BASE_PHYS_ADDR: u64 = 0xfebd1000;

struct LkpEnc {
    _dev: Pin<Box<chrdev::Registration<1>>>,
    // _dev: Pin<Box<miscdev::Registration<LkpEnc>>>,
}

#[vtable]
impl file::Operations for LkpEnc {
    fn open(_context: &Self::OpenData,_file: &file::File) -> Result<Self::Data> {
        pr_info!("qemu file opened\n");
        Ok(())
    }

    fn write(_data: Self::Data,_file: &file::File,reader: &mut impl kernel::io_buffer::IoBufferReader,_offset:u64,) -> Result<usize> {
        pr_info!("qemu file written\n");
        Ok(reader.len())
    }

    fn read(_data: Self::Data,_file: &file::File,writer: &mut impl kernel::io_buffer::IoBufferWriter,offset:u64,) -> Result<usize> {
        pr_info!("qemu file read\n");
    
        // Message to write only once
        let message = b"Hello, World!\n";
        // If the offset is 0, it means we're starting to read from the beginning.
        // If the offset is greater than 0, in this simple case, we assume the message was already read,
        // and thus we return Ok(0) to indicate no more data is to be read.
        if offset == 0 {
            let _ = writer.write_slice(message);
            Ok(message.len())
        } else {
            // No more data to read, indicate this by returning 0 bytes read.
            Ok(0)
        }
    }

    fn ioctl(_data: <Self::Data as kernel::ForeignOwnable>::Borrowed<'_>,_file: &file::File,cmd: &mut file::IoctlCommand,) -> Result<i32> {
        pr_info!("qemu file ioctl\n");
        pr_info!("devmem {:?}", *DEVMEM.lock());
        let io_number = cmd.raw().0;
        // let size = (io_number >> _IOC_SIZESHIFT) & _IOC_SIZEMASK;
        let arg = cmd.raw().1;
        let user_slice = unsafe { UserSlicePtr::new(arg as _, 255 as _)};

        // pr_info!("User slice {:?}", user_slice.read_all());

        match io_number {
            LKP_ENC_WRITE_SEED => {
                let seed = user_slice.read_all().unwrap()[0] as u32;
                pr_info!("LKP_ENC_WRITE_SEED : {:?}", seed);
                unsafe {bindings::iowrite32(seed, *DEVMEM.lock() as _);}
            },
            LKP_ENC_READ_STRING => {
                // pr_info!("LKP_ENC_READ_STRING: {}", unsafe { bindings::ioread32(*DEVMEM.lock() as _) });
                let mut ret: Vec<u8> = Vec::new();
                for i in 0..255 {
                    let c: u8 = unsafe { bindings::ioread8((*DEVMEM.lock() + 4 + i) as _) as u8 };
                    let _ = ret.try_push(c);

                    if c == 0 {
                        break;
                    }              
                }

                pr_info!("LKP_ENC_READ_STRING: {:?}", ret);

                let _ = user_slice.write_all(&ret);
            },
            LKP_ENC_WRITE_STRING => {
                let string = user_slice.read_all();

                match string {
                    Ok(s) => {
                        for (i, c) in s.as_slice().iter().enumerate() {
                            pr_info!("index: {:?}, char: {:?}", i, c);
                            unsafe { bindings::iowrite8(*c, (*DEVMEM.lock() + 4 + i) as _) };
                        }
                    },
                    Err(e) => {
                        pr_info!("Error reading string: {:?}", e);
                    }
                }
                // pr_info!("LKP_ENC_WRITE_STRING: {:?}", string);
                
            },
            _ => {
                pr_info!("Unknown IOCTL command");
            }
        }
        
        Ok(0)
    }
}

impl Module for LkpEnc {
    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
        pr_info!("{} module init", name);
        
        let addr = unsafe { bindings::ioremap(DEVICE_BASE_PHYS_ADDR, 4096) };
        
        // devmem know contains the address mapped in the CPU of the device

        let mut ptr = DEVMEM.lock();
        *ptr = if addr.is_null() {
            Err(ENOMEM)
        } else {
            // INVARIANT: `addr` is non-null and was returned by `ioremap`, so it is valid. It is
            // also 8-byte aligned because we checked it above.
            Ok(addr as usize)
        }?;
        
        // Test the memory mapping and writing to device
        // pr_info!("devmem: {:?}", *ptr);

        // unsafe {
        //     bindings::iowrite32(12, *ptr as _);
        //     pr_info!("devmem: {:?}", bindings::ioread32(*ptr as _));
        // }

        let mut chardev_reg = chrdev::Registration::new_pinned(name, 0, module)?;
        chardev_reg.as_mut().register::<LkpEnc>()?;

        pr_info!("LKP_ENC_WRITE_SEED: {:?}", LKP_ENC_WRITE_SEED);
        pr_info!("LKP_ENC_READ_STRING: {:?}", LKP_ENC_READ_STRING);
        pr_info!("LKP_ENC_WRITE_STRING: {:?}", LKP_ENC_WRITE_STRING);

        Ok(LkpEnc { _dev: chardev_reg })
    }
}

impl Drop for LkpEnc {
    fn drop(&mut self) {
        pr_info!("lkp_enc module exit\n");
        unsafe { bindings::iounmap(*DEVMEM.lock() as *mut c_void ) };
    }
}